src/common/parser.c

```diff
@@ -0,0 +1,399 @@
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * parser.c -- parser of set files
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <ctype.h>
+#include <sys/queue.h>
+
+#include "util.h"
+#include "out.h"
+#include "parser.h"
+
+enum parser_codes {
+	PARSER_CONTINUE = 0,
+	PARSER_PMEMPOOLSET,
+	PARSER_REPLICA,
+	PARSER_SIZE_PATH_EXPECTED,
+	PARSER_WRONG_SIZE,
+	PARSER_WRONG_PATH,
+	PARSER_SET_NO_PARTS,
+	PARSER_REP_NO_PARTS,
+	PARSER_SIZE_MISMATCH,
+	PARSER_FORMAT_OK,
+	PARSER_MAX_CODE
+};
+
+static const char *parser_errstr[PARSER_MAX_CODE] = {
+	"", /* parsing */
+	"the first line must be exactly 'PMEMPOOLSET'",
+	"exactly 'REPLICA' expected",
+	"size and path expected",
+	"incorrect format of size",
+	"incorrect path (must be an absolute path)",
+	"no pool set parts",
+	"no replica parts",
+	"sizes of pool set and replica mismatch",
+	"" /* format correct */
+};
+
+/*
+ * parser_get_next_token -- (internal) extract token from string
+ */
+char *
+parser_get_next_token(char **line)
+{
+	while (*line && isblank(**line))
+		(*line)++;
+
+	return strsep(line, " \t");
+}
+
+/*
+ * parser_read_line -- (internal) read line and validate size and path
+ *                      from a pool set file
+ */
+static enum parser_codes
+parser_read_line(char *line, size_t *size, char **path)
+{
+	char *size_str, *path_str, *endptr;
+
+	size_str = parser_get_next_token(&line);
+	path_str = parser_get_next_token(&line);
+
+	if (!path_str)
+		return PARSER_SIZE_PATH_EXPECTED; /* no path */
+
+	LOG(10, "size '%s' path '%s'", size_str, path_str);
+
+	/*
+	 * A format of the size is checked in detail. As regards the path,
+	 * it is checked only if the read path is an absolute path.
+	 * The rest should be checked during creating/opening the file.
+	 */
+
+	/* check if the read path is an absolute path */
+	if (path_str[0] != '/')
+		return PARSER_WRONG_PATH; /* must be an absolute path */
+
+	*path = Strdup(path_str);
+
+	/* check format of size */
+	*size = strtoull(size_str, &endptr, 10);
+	if (endptr && endptr[0] != '\0') {
+		if (endptr[1] != '\0' || strchr("KMGT", endptr[0]) == NULL)
```

Comment:
Why are the suffix characters required?  Can't I just put the number of bytes I want for the size?